src/ui/src/mainwindow.cpp

```diff
@@ -1,44 +1,101 @@
 #include "mainwindow.h"
 
 #include <QFileDialog>
+#include <QMessageBox>
 
+#include "log.h"
+#include "scene.h"
 #include "ui_mainwindow.h"
 
 MainWindow::MainWindow(QWidget* parent)
-		: QMainWindow(parent), ui_(new Ui::MainWindow) {
+		: QMainWindow(parent),
+			scene_(std::make_shared<Scene>()),
+			ui_(new Ui::MainWindow) {
 	ui_->setupUi(this);
 	setWindowTitle(tr("OVERHEAT application"));
 	// renderer widget
-	render_widget_ = new RendererWidget(this);
+	render_widget_ = new RendererWidget(this, scene_);
 	setCentralWidget(render_widget_);
 	render_widget_->showMaximized();
 	// shape widget
 	shape_list_widget_ = new ShapeListWidget(this);
 	addDockWidget(Qt::LeftDockWidgetArea, shape_list_widget_);
 	shape_list_widget_->setWindowTitle("Shapes");
 	// metadata widget
-	metadata_widget_ = new MetadataWidget(this);
-	addDockWidget(Qt::LeftDockWidgetArea, metadata_widget_);
-	metadata_widget_->setWindowTitle("Metadata");
+	metadata_widget_           = new MetadataWidget(this);
+	QDockWidget* metadata_dock = new QDockWidget(tr("Metadata"), this);
+	metadata_dock->setWidget(metadata_widget_);
+	addDockWidget(Qt::LeftDockWidgetArea, metadata_dock);
+	connect(this, &MainWindow::ShowMetadata, metadata_widget_,
+	        &MetadataWidget::OnShowMetadata);
+
 	// options widget
 	visualization_options_ = new VisualizationOptionsWidget(this);
 	addDockWidget(Qt::LeftDockWidgetArea, visualization_options_);
 	visualization_options_->setWindowTitle("Options");
 	// signals and slots connection
 	connect(ui_->load_file_btn, &QAction::triggered, this,
 	        &MainWindow::OnLoadFileBtnPressed);
+	connect(visualization_options_,
+	        &VisualizationOptionsWidget::VisualizationOptionsChanged,
+	        render_widget_, &RendererWidget::onVisualizationOptionsChanged);
```

Comment:
Please, initialize members of class in initializer list. Beside of perfomance gains it saves you from using them uninitialized in ctor body